File: apps/desktop/src/main/kotlin/store/ChatStore.kt
package store

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ChatStore {
    private val storeScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val typingTimers = java.util.concurrent.ConcurrentHashMap<Pair<String, String>, Job>()
    private val typingTimeoutMs = 5_000L

    private val _chats = MutableStateFlow<List<ChatItem>>(emptyList())
    val chats: StateFlow<List<ChatItem>> = _chats.asStateFlow()

    private val _messages = MutableStateFlow<Map<String, List<MessageItem>>>(emptyMap())
    val messages: StateFlow<Map<String, List<MessageItem>>> = _messages.asStateFlow()

    private val _typing = MutableStateFlow<Map<String, Set<String>>>(emptyMap())
    val typing: StateFlow<Map<String, Set<String>>> = _typing.asStateFlow()

    private val _call = MutableStateFlow(CallState())
    val call: StateFlow<CallState> = _call.asStateFlow()

    private val _callError = MutableStateFlow<String?>(null)
    val callError: StateFlow<String?> = _callError.asStateFlow()

    fun setCallError(message: String) { _callError.value = message }
    fun clearCallError() { _callError.value = null }

    fun setChats(list: List<ChatItem>) { _chats.value = list }

    fun isGroup(chatId: String): Boolean =
        _chats.value.find { it.id == chatId }?.isGroup ?: false

    fun onMessageReceived(chatId: String, clientMsgId: String, plaintext: String, senderId: String, timestamp: Long) {
        val msg = MessageItem(
            id = clientMsgId, clientMsgId = clientMsgId, chatId = chatId,
            senderId = senderId, plaintext = plaintext, timestamp = timestamp,
            status = "delivered", isDeleted = false,
        )
        val current = _messages.value.toMutableMap()
        current[chatId] = (current[chatId] ?: emptyList()) + msg
        _messages.value = current

        val chats = _chats.value.toMutableList()
        val idx = chats.indexOfFirst { it.id == chatId }
        if (idx >= 0) {
            chats[idx] = chats[idx].copy(lastMessage = plaintext, updatedAt = timestamp)
            _chats.value = chats.sortedByDescending { it.updatedAt }
        }
    }

    fun onMessageStatusUpdate(clientMsgId: String, status: String) {
        val updated = _messages.value.mapValues { (_, msgs) ->
            msgs.map { if (it.clientMsgId == clientMsgId) it.copy(status = status) else it }
        }
        _messages.value = updated
    }

    fun onTyping(chatId: String, userId: String) {
        val current = _typing.value.toMutableMap()
        current[chatId] = (current[chatId] ?: emptySet()) + userId
        _typing.value = current

        val key = chatId to userId
        typingTimers[key]?.cancel()
        typingTimers[key] = storeScope.launch {
            delay(typingTimeoutMs)
            onTypingStop(chatId, userId)
        }
    }

    fun onTypingStop(chatId: String, userId: String) {
        val current = _typing.value.toMutableMap()
        val updated = (current[chatId] ?: emptySet()) - userId
        if (updated.isEmpty()) current.remove(chatId) else current[chatId] = updated
        _typing.value = current
        typingTimers.remove(chatId to userId)?.cancel()
    }

    fun onRead(chatId: String, messageId: String) {
        onMessageStatusUpdate(messageId, "read")
    }

    fun onMessageDeleted(clientMsgId: String) {
        val updated = _messages.value.mapValues { (_, msgs) ->
            msgs.map { if (it.clientMsgId == clientMsgId) it.copy(isDeleted = true) else it }
                .filter { !it.isDeleted }
        }
        _messages.value = updated
    }

    fun onMessageEdited(clientMsgId: String, newPlaintext: String) {
        val updated = _messages.value.mapValues { (_, msgs) ->
            msgs.map { if (it.clientMsgId == clientMsgId) it.copy(plaintext = newPlaintext) else it }
        }
        _messages.value = updated
    }

    fun setMessages(chatId: String, msgs: List<MessageItem>) {
        val current = _messages.value.toMutableMap()
        current[chatId] = msgs
        _messages.value = current
    }

    fun onCallOffer(callId: String, chatId: String, fromUserId: String, isVideo: Boolean = false) {
        _call.value = CallState(CallStatus.RINGING_IN, callId, chatId, fromUserId, isVideo)
    }

    fun markLocalVideoReady(callId: String) {
        if (_call.value.callId == callId) _call.value = _call.value.copy(hasLocalVideo = true)
    }

    fun markRemoteVideoReady(callId: String) {
        if (_call.value.callId == callId) _call.value = _call.value.copy(hasRemoteVideo = true)
    }

    fun onCallAnswer(callId: String) {
        val cur = _call.value
        if (cur.callId == callId) _call.value = cur.copy(status = CallStatus.ACTIVE)
    }

    fun onCallEnd(callId: String) {
        if (_call.value.callId == callId) _call.value = CallState()
    }

    fun setOutgoingCall(callId: String, chatId: String, targetId: String, isVideo: Boolean) {
        _call.value = CallState(CallStatus.RINGING_OUT, callId, chatId, targetId, isVideo)
    }

    fun clearCall() { _call.value = CallState() }
}

File: apps/mobile/ios/Sources/Messenger/store/ChatStore.swift
// ChatStore.swift — реактивное хранилище состояния чатов.
// Зеркало ChatStore.kt (Android).
// @MainActor гарантирует обновление @Published только на главном потоке.

import Foundation
import Combine

@MainActor
final class ChatStore: ObservableObject {

    @Published private(set) var chats: [ChatItem] = []
    @Published private(set) var messages: [String: [MessageItem]] = [:]
    @Published private(set) var typing: [String: Set<String>] = [:]
    @Published private(set) var callState = CallState()
    @Published var callError: String? = nil

    func setCallError(_ message: String) { callError = message }
    func clearCallError() { callError = nil }

    // MARK: - Chats

    func setChats(_ list: [ChatItem]) { chats = list }

    func isGroup(_ chatId: String) -> Bool {
        chats.first(where: { $0.id == chatId })?.isGroup ?? false
    }

    // MARK: - Messages

    func setMessages(chatId: String, msgs: [MessageItem]) {
        messages[chatId] = msgs
    }

    func addMessage(chatId: String, item: MessageItem) {
        var list = messages[chatId] ?? []
        list.append(item)
        list.sort { $0.timestamp < $1.timestamp }
        messages[chatId] = list
    }

    func onMessageReceived(chatId: String, clientMsgId: String, plaintext: String,
                           senderId: String, timestamp: Int64) {
        let msg = MessageItem(
            id: clientMsgId, clientMsgId: clientMsgId, chatId: chatId,
            senderId: senderId, plaintext: plaintext, timestamp: timestamp,
            status: "delivered", isDeleted: false
        )
        addMessage(chatId: chatId, item: msg)

        // Обновить последнее сообщение в списке чатов
        if let idx = chats.firstIndex(where: { $0.id == chatId }) {
            var chat = chats[idx]
            chat.lastMessage = plaintext
            chat.updatedAt = timestamp
            chats[idx] = chat
            chats.sort { $0.updatedAt > $1.updatedAt }
        }
    }

    func onMessageStatusUpdate(clientMsgId: String, status: String) {
        messages = messages.mapValues { msgs in
            msgs.map { $0.clientMsgId == clientMsgId ? MessageItem(
                id: $0.id, clientMsgId: $0.clientMsgId, chatId: $0.chatId,
                senderId: $0.senderId, plaintext: $0.plaintext, timestamp: $0.timestamp,
                status: status, isDeleted: $0.isDeleted,
                mediaId: $0.mediaId, mediaKey: $0.mediaKey,
                originalName: $0.originalName, contentType: $0.contentType
            ) : $0 }
        }
    }

    func onMessageDeleted(clientMsgId: String) {
        messages = messages.mapValues { msgs in
            msgs.filter { $0.clientMsgId != clientMsgId }
        }
    }

    func onMessageEdited(clientMsgId: String, newPlaintext: String) {
        messages = messages.mapValues { msgs in
            msgs.map { m in
                guard m.clientMsgId == clientMsgId else { return m }
                return MessageItem(
                    id: m.id, clientMsgId: m.clientMsgId, chatId: m.chatId,
                    senderId: m.senderId, plaintext: newPlaintext, timestamp: m.timestamp,
                    status: m.status, isDeleted: m.isDeleted,
                    mediaId: m.mediaId, mediaKey: m.mediaKey,
                    originalName: m.originalName, contentType: m.contentType
                )
            }
        }
    }

    // MARK: - Typing

    func onTyping(chatId: String, userId: String) {
        var t = typing[chatId] ?? []
        t.insert(userId)
        typing[chatId] = t
    }

    func onTypingStop(chatId: String, userId: String) {
        var t = typing[chatId] ?? []
        t.remove(userId)
        typing[chatId] = t.isEmpty ? nil : t
    }

    // MARK: - Calls

    func onCallOffer(callId: String, chatId: String, fromUserId: String, isVideo: Bool = false) {
        callState = CallState(status: .ringingIn, callId: callId, chatId: chatId,
                              remoteUserId: fromUserId, isVideo: isVideo)
    }

    func onCallAnswer(callId: String) {
        guard callState.callId == callId else { return }
        callState = CallState(status: .active, callId: callId, chatId: callState.chatId,
                              remoteUserId: callState.remoteUserId, isVideo: callState.isVideo)
    }

    func setOutgoingCall(callId: String, chatId: String, targetId: String, isVideo: Bool) {
        callState = CallState(status: .ringingOut, callId: callId, chatId: chatId,
                              remoteUserId: targetId, isVideo: isVideo)
    }

    func onCallEnd(callId: String) {
        if callState.callId == callId { callState = CallState() }
    }

    func markLocalVideoReady()  { callState.hasLocalVideo  = true }
    func markRemoteVideoReady() { callState.hasRemoteVideo = true }

    func clearCall() { callState = CallState() }

    func onRead(chatId: String, messageId: String) {
        onMessageStatusUpdate(clientMsgId: messageId, status: "read")
    }
}
